File: Berger.Data/MsfaEntity/DealerSalesCall/DealerSalesCall.cs
using Berger.Common.Enumerations;
using Berger.Data.Common;
using Berger.Data.MsfaEntity.PainterRegistration;
using Berger.Data.MsfaEntity.SAPTables;
using Berger.Data.MsfaEntity.Setup;
using Berger.Data.MsfaEntity.Users;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Berger.Data.MsfaEntity.DealerSalesCall
{
    public class DealerSalesCall : AuditableEntity<int>
    {
        public int DealerId { get; set; }
        public DealerInfo Dealer { get; set; }
        public int UserId { get; set; }
        public UserInfo User { get; set; }
        public int? JourneyPlanId { get; set; }
        public JourneyPlanMaster JourneyPlan { get; set; }
        //public DateTime Date { get; set; }
        public bool IsTargetPromotionCommunicated { get; set; }
        public bool IsTargetCommunicated { get; set; }

        //public EnumRatings SecondarySalesRatings { get; set; }
        public int SecondarySalesRatingsId { get; set; }
        public DropdownDetail SecondarySalesRatings { get; set; }
        public string SecondarySalesReasonTitle { get; set; }
        public string SecondarySalesReasonRemarks { get; set; }

        public bool HasOS { get; set; }
        public bool IsOSCommunicated { get; set; }
        public bool HasSlippage { get; set; }
        public bool IsSlippageCommunicated { get; set; }

        public bool IsPremiumProductCommunicated { get; set; }
        public bool IsPremiumProductLifting { get; set; }
        //public EnumProductLifting? PremiumProductLifting { get; set; }
        public int? PremiumProductLiftingId { get; set; }
        public DropdownDetail PremiumProductLifting { get; set; }
        public string PremiumProductLiftingOthers { get; set; }

        public bool IsCBInstalled { get; set; }
        public bool IsCBProductivityCommunicated { get; set; }

        //public bool IsMerchendisingPlanogramFollowed { get; set; }
        public int? MerchendisingId { get; set; }
        public DropdownDetail Merchendising { get; set; }

        public bool HasSubDealerInfluence { get; set; }
        //public EnumSubDealerInfluence? SubDealerInfluence { get; set; }
        public int? SubDealerInfluenceId { get; set; }
        public DropdownDetail SubDealerInfluence { get; set; }

        public bool HasPainterInfluence { get; set; }
        //public EnumPainterInfluence? PainterInfluence { get; set; }
        public int? PainterInfluenceId { get; set; }
        public DropdownDetail PainterInfluence { get; set; }

        public bool IsShopManProductKnowledgeDiscussed { get; set; }
        public bool IsShopManSalesTechniquesDiscussed { get; set; }
        public bool IsShopManMerchendizingImprovementDiscussed { get; set; }

        public bool HasCompetitionPresence { get; set; }
        //public EnumCompetitionPresence? CompetitionPresence { get; set; }
        public bool IsCompetitionServiceBetterThanBPBL { get; set; }
        public string CompetitionServiceBetterThanBPBLRemarks { get; set; }
        public bool IsCompetitionProductDisplayBetterThanBPBL { get; set; }
        public string CompetitionProductDisplayBetterThanBPBLRemarks { get; set; }
        //public int? CompetitionProductDisplayImageId { get; set; }
        //public Attachment CompetitionProductDisplayImage { get; set; }
        public string CompetitionProductDisplayImageUrl { get; set; }
        public string CompetitionSchemeModalityComments { get; set; }
        //public int? CompetitionSchemeModalityImageId { get; set; }
        //public Attachment CompetitionSchemeModalityImage { get; set; }
        public string CompetitionSchemeModalityImageUrl { get; set; }
        public string CompetitionShopBoysComments { get; set; }
        public IList<DealerCompetitionSales> DealerCompetitionSales { get; set; }

        public bool HasDealerSalesIssue { get; set; }
        //public EnumDealerSalesIssue? EnumDealerSalesIssue { get; set; }
        public IList<DealerSalesIssue> DealerSalesIssues { get; set; }

        //public EnumSatisfaction DealerSatisfaction { get; set; }
        public int DealerSatisfactionId { get; set; }
        public DropdownDetail DealerSatisfaction { get; set; }
        public string DealerSatisfactionReason { get; set; }

        // for sub dealer
        public bool IsSubDealerCall { get; set; }
        public bool HasBPBLSales { get; set; }
        public decimal BPBLAverageMonthlySales { get; set; }
        public decimal BPBLActualMTDSales { get; set; }
    }
}

File: BergerMsfaApi/ClientApp/src/app/Shared/Entity/DealerSalesCall/dealer-sales-call.ts
import { QueryObject } from '../Common/query-object';
import { Dropdown } from '../Setup/dropdown';
import { UserInfo } from '../Users';

export class DealerSalesCall {
    id: number;
    dealerId: number;
    // dealer: any;
    dealerName: string;
    userId: number;
    // user: UserInfo;
    userFullName: string;
    journeyPlanId: number;
    date: Date;
    dateText: string;
    isTargetPromotionCommunicated: boolean;
    isTargetCommunicated: boolean;

    secondarySalesRatingsId: number;
    // secondarySalesRatings: Dropdown;
    secondarySalesRatingsText: string;
    secondarySalesReasonTitle: string;
    secondarySalesReasonRemarks: string;

    hasOS: boolean;
    isOSCommunicated: boolean;
    hasSlippage: boolean;
    isSlippageCommunicated: boolean;

    isPremiumProductCommunicated: boolean;
    isPremiumProductLifting: boolean;
    premiumProductLiftingId: number;
    // premiumProductLifting: Dropdown;
    premiumProductLiftingText: string;
    premiumProductLiftingOthers: string;

    isCBInstalled: boolean;
    isCBProductivityCommunicated: boolean;

    merchendisingId: number;
    // merchendising: Dropdown;
    merchendisingText: string;

    hasSubDealerInfluence: boolean;
    subDealerInfluenceId: number;
    // subDealerInfluence: Dropdown;
    subDealerInfluenceText: string;

    hasPainterInfluence: boolean;
    painterInfluenceId: number;
    // painterInfluence: Dropdown;
    painterInfluenceText: string;

    isShopManProductKnowledgeDiscussed: boolean;
    isShopManSalesTechniquesDiscussed: boolean;
    isShopManMerchendizingImprovementDiscussed: boolean;

    hasCompetitionPresence: boolean;
    isCompetitionServiceBetterThanBPBL: boolean;
    competitionServiceBetterThanBPBLRemarks: string;
    isCompetitionProductDisplayBetterThanBPBL: boolean;
    competitionProductDisplayBetterThanBPBLRemarks: string;
    competitionProductDisplayImageUrl: string;
    competitionSchemeModalityComments: string;
    competitionSchemeModalityImageUrl: string;
    competitionShopBoysComments: string;
    dealerCompetitionSales: DealerCompetitionSales[];

    hasDealerSalesIssue: boolean;
    dealerSalesIssues: DealerSalesIssue[];

    dealerSatisfactionId: number;
    // dealerSatisfaction: Dropdown;
    dealerSatisfactionText: string;
    dealerSatisfactionReason: string;

    // for sub dealer
    isSubDealerCall: boolean;
    hasBPBLSales: boolean;
    bPBLAverageMonthlySales: number;
    bPBLActualMTDSales: number;

    // display text
    detailsBtnText: string;
    // isTargetPromotionCommunicatedText: string;
    // isTargetCommunicatedText: string;
    // hasOSText: string;
    // isOSCommunicatedText: string;
    // hasSlippageText: string;
    // isSlippageCommunicatedText: string;
    // isPremiumProductCommunicatedText: string;
    // isPremiumProductLiftingText: string;
    // isCBInstalledText: string;
    // isCBProductivityCommunicatedText: string;
    // hasSubDealerInfluenceText: string;
    // hasPainterInfluenceText: string;
    // isShopManProductKnowledgeDiscussedText: string;
    // isShopManSalesTechniquesDiscussedText: string;
    // isShopManMerchendizingImprovementDiscussedText: string;
    // hasCompetitionPresenceText: string;
    // isCompetitionServiceBetterThanBPBLText: string;
    // isCompetitionProductDisplayBetterThanBPBLText: string;
    // hasDealerSalesIssueText: string;
    // isSubDealerCallText: string;
    // hasBPBLSalesText: string;

    // [key: string]: any;

    constructor(init?: Partial<DealerSalesCall>) {
        Object.assign(this, init);
    }

    clear() {
        this.id = null;
        // this.isTargetPromotionCommunicated = false;
        // this.isTargetCommunicated = false;
        // this.hasOS = false;
        // this.isOSCommunicated = false;
        // this.hasSlippage = false;
        // this.isSlippageCommunicated = false;
        // this.isPremiumProductCommunicated = false;
        // this.isPremiumProductLifting = false;
        // this.isCBInstalled = false;
        // this.isCBProductivityCommunicated = false;
        // this.hasSubDealerInfluence = false;
        // this.hasPainterInfluence = false;
        // this.isShopManProductKnowledgeDiscussed = false;
        // this.isShopManSalesTechniquesDiscussed = false;
        // this.isShopManMerchendizingImprovementDiscussed = false;
        // this.hasCompetitionPresence = false;
        // this.isCompetitionServiceBetterThanBPBL = false;
        // this.isCompetitionProductDisplayBetterThanBPBL = false;
        // this.hasDealerSalesIssue = false;
        // this.isSubDealerCall = false;
        // this.hasBPBLSales = false;
    }

    // booleanToYesNoText() {
    //     let entries = Object.entries(this) || [];
    //     console.log(entries);
    //     entries.forEach(([key, value]) => {
    //         // let value = this[key];
    //         let keyText = key+'Text';
    //         if (typeof value === "boolean") {
    //             this[keyText] = value ? 'YES' : 'NO'; 
    //         }
    //     });
    // }
}

export class DealerCompetitionSales {
    dealerSalesCallId: number;
    //dealerSalesCall: DealerSalesCall;
    companyId: number;
    // company: Dropdown;
    companyName: string;
    averageMonthlySales: number;
    actualMTDSales: number;
    
    constructor(init?: Partial<DealerCompetitionSales>) {
        Object.assign(this, init);
    }

    clear() {
    }
}

export class DealerSalesIssue {
    dealerSalesCallId: number;
    //dealerSalesCall: DealerSalesCall;
    dealerSalesIssueCategoryId: number;
    // dealerSalesIssueCategory: Dropdown;
    dealerSalesIssueCategoryText: string;
    materialName: string;
    materialGroup: string;
    quantity: number;
    batchNumber: string;
    comments: string;
    priorityId: number;
    // priority: Dropdown;
    priorityText: string;

    hasCBMachineMantainance: number;
    cBMachineMantainanceId: number;
    // cBMachineMantainance: Dropdown;
    cBMachineMantainanceText: string;
    cBMachineMantainanceRegularReason: string;

    constructor(init?: Partial<DealerSalesIssue>) {
        Object.assign(this, init);
    }

    clear() {
    }
}

export class DealerSalesCallQuery extends QueryObject {
    title: string;
    
    constructor(init?: Partial<DealerSalesCallQuery>) {
        super();
        Object.assign(this, init);
    }

    clear() {
        this.title = '';
    }
}